tune-threads/tune-threads.lisp

```diff
@@ -0,0 +1,169 @@
+    (format t "~d: ~a~%" (optimal-num-threads-from-scan tm) (scaled-times-per-trial tm))))
+
+(defun time-per-trial (tm)
+  "Compute number of seconds per thread, per trial from timing data in instance of TIMING struct TM."
+  (/ (timing-run-time tm) (* (timing-num-threads tm) (timing-num-trials tm))))
+
+(defun times-per-trial (tms)
+  "Compute timing via TIME-PER-TRIAL for each entry in hash table TMS."
+  (map 'vector (lambda (num-threads) (time-per-trial (gethash num-threads tms)))
+       (sort-hash-table-keys tms)))
+
+(defun scaled-times-per-trial (tms)
+  "Compute and collect the timing from each entry in hash table TMS and rescale by the minimum time.
+A vector is returned reprenting the relative efficiecy for a fixed number of qubits and varying number of threads."
+  (rescale-by-minimum (times-per-trial tms)))
+
+(defun optimal-num-threads-from-scan (tms)
+  "Compute the optimal number of threads from the hash table of TIMING structs TMS
+whose keys are different numbers of threads."
+  (let* ((run-times (times-per-trial tms))
+         (min-index (findmin run-times)))
+    (elt (sort-hash-table-keys tms) min-index)))
+
+(defparameter *default-target-time* 1.0
+  "The approximate time consumed by a block of repeated trials during precise timing.")
+
+(defun optimal-num-threads (num-qubits
+                            &key (target-time *default-target-time*) (num-threads-to-scan nil)
+                                 (prepare-qvm-program #'prepare-hadamard-test) (print-info nil))
+  "Compute the optimal number of threads for a NUM-QUBITS qubit program. Keys are:
+TARGET-TIME -- approximate seconds to be consumed in precise timing. Default 1s.
+NUM-THREADS-TO-SCAN -- a list whose elements are different numbers of threads to time.
+                       Defaults to nil which means one through the number of cores.
+PREPARE-QVM-PROGRAM -- a function that takes one argument, the number of qubits, and returns a compiled quil program.
+PRINT-INFO -- print very brief summary of timing if true."
+  (optimal-num-threads-from-scan (scan-num-threads num-qubits
+                                         :prepare-qvm-program prepare-qvm-program :print-info print-info
+                                         :num-threads-to-scan num-threads-to-scan
+                                         :target-time target-time)))
+
+(defun time-quil-prog (q num-threads target-time &key (init-num-pre-trials 1))
+  "Time the quil program in qvm Q and return a TIMING struct. NUM-THREADS threads are used during timing.
+Approximately TARGET-TIME seconds will be consumed during timing.
+The keyword parameter :INIT-NUM-PRE-TRIALS determines the intial number of trials to use during pre-timing.
+Pre-timing is used to choose the number of trials to consume TARGET-TIME seconds."
+  (let* ((pt (run-pre-timing q num-threads init-num-pre-trials))
+         (num-trials (num-trials-in-target-time pt target-time))
```

Comment:
As long as I'm throwing out simple-minded questions, [here comes another one! (warning: Monty Python Youtube link)](https://youtu.be/zpv5kFMcq8I)

What would be the downside to just picking a fixed num-trials for a given num-qubits, then just measuring wall-clock time for each value of num-threads and comparing them at the end?

I ask because I have a vague worry about the fact that we recalculate num-trials for each new value of num-threads. But I admit my thinking here is unclear, and I don't have any strong argument for why it should matter. It feels like we are trying to determine which value of num-threads minimizes run time, but we are "changing the experiment" each time, then attempting to re-adjust the results after the fact by taking num-trials into account.

Note that if you fixed num-trials in `scan-num-threads`, you could still do a "pre-timing" pass to pick num-trials such that `num-trials * fastest-pre-timing-run-per-trial ~= time-limit`. This would make time-limit a rough lower bound on "run-time per num-threads experiment", rather than an upper bound.

The fixed-num-trials approach has the obvious downside of taking longer, since you no longer cap each experiment to ~ time-limit seconds, but has the advantage of being more of an "apples-to-apples" comparison between the runtimes since you are only varying num-threads rather than num-threads and num-trials simultaneously.

I admit that the main reason I like this fixed-num-trials scheme is that it is simpler to reason about, and I am a simpleton and hence like simple things :-). You should feel free to ignore this if it's clear to you that changing num-trials between runs does not invalidate the timing experiments.

Reply:
You know, my life is a lot easier if no one pays attention to what I do... code quality suffers a bit though ;)  Short answer. Yes, that's a good idea. But, for the situations I have tested, the results are rather robust to these details.  I don't know if that will always be the case for realistic scenarios, eg. using a different quil program. But, I hope the results are not so sensitive. On the other hand, with a lot of cores, say 24, you may want to adjust the number of trials so you don't have to wait as long.

btw. In the current iteration, you may have noticed that pre-trials starting with 1 trial are run for every number of qubits and every number of threads. This is ok, until I test large numbers of qubits, and test this code at the same time (i.e. I need to do tests repeatedly while changing the code) In this case, I always run one pretrial and then one trial. I have added a feature (not pushed) to specify the number of trials and at the same time skip pretrials. I don't like the way it's implemented at the moment.

One reason that using previous results to set the number of pretrials is not implemented is because the numbers of qubits to try is now specified in a list. If all numbers of qubits are tried sequentially, then using half the final number of pretrials as the initial number of pretrials for the next test may be ok. But, if someone asks for 4 qubits and then 24, the test will never finish.

Reply:
I remember the contractual obligation record pretty well. Haven't heard it in more years than I care to admit.

Reply:
> I remember the contractual obligation record pretty well. Haven't heard it in more years than I care to admit.

I think I had that record on vinyl, picked up from a discount library sale, but I might be mis-remembering that.  I'll see if I can work in a "Finland... Finland... Finland..." joke next time.

I was not considering the case of a many-core server, or your testing travails, both very good reasons to want to speed up the scanning.

Reply:
I added an option to fix the number of trials while scanning number of threads. Its useful to understand what's happening. But, the time per trial can vary by more than an order of magnitude.

Reply:
Thanks for humoring me :).  I guess the order-of-magnitude runtime differences makes it impractical for scanning with large upper bounds on num-threads and num-qubits, but is also a good sign that being able to tune-threads is a very useful thing to do.